### PR TITLE
types: implement new Restore func of ValueExpr

### PR DESCRIPTION
**Pull request by leoppro**
### What problem does this PR solve? 
 implement new Restore func of ValueExpr

See: pingcap/parser#78

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8679)

### CODE REVIEW

**Comment by leoppro** [comment]
@XuHuaiyu @tiancaiamao PTAL

File: types/parser_driver/value_expr.go [hunk 1]
```diff
@@ -69,11 +68,39 @@ type ValueExpr struct {
...
+		ctx.WritePlain(strconv.FormatFloat(n.GetFloat64(), 'e', -1, 32))
+	case types.KindFloat64:
+		ctx.WritePlain(strconv.FormatFloat(n.GetFloat64(), 'e', -1, 64))
+	case types.KindString, types.KindBytes:
+		ctx.WriteString(n.GetString())
+	case types.KindMysqlDecimal:
+		ctx.WritePlain(n.GetMysqlDecimal().String())
+	case types.KindBinaryLiteral:
+		if n.Type.Flag&mysql.UnsignedFlag != 0 {
+			_, _ = fmt.Fprintf(ctx.In, "x'%x'", n.GetBytes())
```
Comment (kennytm) [inline types/parser_driver/value_expr.go]: IMO it is better to 

```go
ctx.WritePlain(fmt.Sprintf("x'%x'", n.GetBytes()))
```

or expose a new function

```go
ctx.WriteFormat("x'%x'", n.GetBytes())
```

than directly using the field `ctx.In`.

**Comment by leoppro** [comment]
> Can we add a test for this?

OK @XuHuaiyu

File: types/parser_driver/value_expr.go [hunk 2]
```diff
@@ -69,11 +68,45 @@ type ValueExpr struct {
...
-	err := n.format(sb)
-	if err != nil {
-		return errors.Trace(err)
+// Restore implements Node interface.
+func (n *ValueExpr) Restore(ctx *ast.RestoreCtx) error {
+	switch n.Kind() {
+	case types.KindNull:
+		ctx.WriteKeyWord("NULL")
+	case types.KindInt64:
+		if n.Type.Flag&mysql.IsBooleanFlag != 0 {
```
Comment (leoppro) [inline types/parser_driver/value_expr.go]: what is `mysql.IsBooleanFlag` ? i copy it from format.
```
if n.Type.Flag&mysql.IsBooleanFlag != 0 {
			if n.GetInt64() > 0 {
				ctx.WriteKeyWord("TRUE")
			} else {
				ctx.WriteKeyWord("FALSE")
			}
}
```
is it right?

**Comment by leoppro** [comment]
@XuHuaiyu @kennytm  PTAL

File: go.mod [hunk 3]
```diff
@@ -47,7 +47,7 @@ require (
 	github.com/pingcap/errors v0.11.0
 	github.com/pingcap/goleveldb v0.0.0-20171020122428-b9ff6c35079e
 	github.com/pingcap/kvproto v0.0.0-20181105061835-1b5d69cd1d26
-	github.com/pingcap/parser v0.0.0-20181211162423-79b42ecfaf70
+	github.com/pingcap/parser v0.0.0-20181214132045-732efe993f70
```
Comment (tiancaiamao) [inline go.mod]:
```
➜  tidb git:(master) ✗ GO111MODULE=on go get -u -v github.com/pingcap/parser@732efe993f70da99fdc18acb380737be33f2333a
➜  tidb git:(master) ✗ GO111MODULE=on go mod tidy
```

Then `make dev` on this branch, make sure the all test passes.
If the CI still fail, I suspect that something is wrong with the downloaded mod cache

Comment (leoppro) [reply go.mod]: see pingcap/tidb#8694 please, i only changed a proposal, but the ci also fail

Comment (tiancaiamao) [reply go.mod]: `make dev` success? so it must be the later case @leoppro

File: types/parser_driver/value_expr.go (hunk 2 above)
Comment (tiancaiamao) [reply types/parser_driver/value_expr.go]: It's right. boolean is represent in integer.

**Pull request merged**